frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py

```diff
@@ -0,0 +1,472 @@
+    super().__init__()
+    self.reg_ops = reg_ops
+    self.ods_emitter = ods_emitter
+    self.impl_emitter = impl_emitter
+
+  def print_banner(self, text):
+    for em in (self.ods_emitter, self.impl_emitter):
+      em.print(
+          "// -----------------------------------------------------------------------------"
+      )
+      em.print(f"// {text}")
+      em.print(
+          "// -----------------------------------------------------------------------------"
+      )
+      em.print("")
+
+  def std_binary_op(self, reg_name, ods_name, op_name):
+    """"Binary"-ops. These ops typically have:
+      - '.Tensor' variant where the second arg is a Tensor
+      - '.Scalar' variant where the second arg is a Scalar
+      - An '.out' variant which contains a final "outref" argument
+    Actual suffixes vary and the rules are set up to match anything
+    that comes in in one of these forms.
+    In addition, most of these have in-place versions, possibly of the
+    '.Tensor' and '.Scalar' variants above.
+    Note that many of these have more than two arguments (i.e. alpha/beta
+    scalars trailing and such), but this is not load bearing.
+    We generally canonicalize all of these forms to a single recognized op
+    by:
+      - Enabling the flag to promotTrailingOutTensor
+      - Enabling the flag matchInplaceVariant
+      - Setting all arguments and returns to kImmutableTensor
+      - Enabling kPromoteScalarToTensor on the second argument.
+    """
+    reg_record = self._get_reg_record(reg_name, f"{reg_name}.Tensor")
+    ods_ins, arg_type_flags = self._map_sigtypes(
+        reg_record["arguments"],
+        type_transforms={
+            "Tensor:0": "AnyTorchImmutableTensor",
+            "Tensor:1": "AnyTorchImmutableTensor",
+            "Scalar:1": "AnyTorchImmutableTensor",
+            "Scalar": "AnyTorchScalarType",
+        },
+        flag_transforms={
+            ":0": ["kImmutableTensor"],
+            ":1": ["kImmutableTensor", "kPromoteScalar"],
+        })
+    ods_outs, return_type_flags = self._map_sigtypes(
+        reg_record["returns"],
+        type_transforms={
+            "Tensor:0": "AnyTorchImmutableTensor",
+        },
+        flag_transforms={
+            ":0": ["kImmutableTensor"],
+        })
+    self.ods_emitter.emit_opdef(ods_name,
+                                op_name,
+                                reg_record,
+                                ods_ins=ods_ins,
+                                ods_outs=ods_outs,
+                                traits=["NoSideEffect"])
+    self.impl_emitter.emit_kernel_methods(ods_name,
+                                          reg_record,
+                                          arg_type_flags=arg_type_flags,
+                                          return_type_flags=return_type_flags,
+                                          promote_trailing_out_tensor=True)
+
+  def std_unary_op(self, reg_name, ods_name, op_name):
```

Comment:
"std" here has any specific meaning. could we use less overloaded terminology like "ordinary"? Or maybe just call it `binary_op`?

Reply:
Switched to 'ordinary'. There are variations and they will become important later.